perf: reduce calls to GetWorkspaceByAgentID in GetWorkspaceAgentByID

cstyan opened the pull request:
This PR piggy backs on the agent API cached workspace added in https://github.com/coder/coder/pull/20662 to provide a fast path for avoiding `GetWorkspaceByAgentID` calls in dbauthz's `GetWorkspaceAgentByID`. This query is not the most expensive, but has a significant call volume at ~16 million calls per week. 

Profiling shows that the majority of these calls are via the agent API's `agentFn`. 
<img width="1688" height="210" alt="Screenshot 2025-12-01 at 2 37 21 PM" src="https://github.com/user-attachments/assets/167aaef5-8e5e-406e-a8f2-c2da21b330ef" />

zedkipp commented on coderd/agentapi/stats.go:
```diff
@@ -43,7 +44,20 @@ func (a *StatsAPI) UpdateStats(ctx context.Context, req *agentproto.UpdateStatsR
 
-	workspaceAgent, err := a.AgentFn(ctx)
+	// Inject RBAC object into context for dbauthz fast path, avoid having to
+	// call GetWorkspaceByAgentID on every metadata update.
```
comment copy-pasta? I think this is updating stats rather than metadata.

cstyan commented on coderd/agentapi/connectionlog_test.go:
it's not very easy to do with the amount of mocking we do in some of these tests, especially given that tests circumvent actually calls to the DB to retrieve the agent in the `agentFn`

the function and test that are updated in `dbauthz` are the code path that will be executed in production when `agentFn` is called from any of the sub-APIs in the agentapi package

Emyrk reviewed (approved):
The helper function to deduplicate code would be preferred :+1:

cstyan committed:
perf: support fastpath in dbauthz GetLatestWorkspaceBuildByWorkspaceID (#21047)

This PR piggy backs on the agent API cached workspace added in earlier PRs to provide a fast path for avoiding `GetWorkspaceByID` calls in `GetLatestWorkspaceBuildByWorkspaceID` via injection of the workspaces RBAC object into the context. We can do this from the `agentConnectionMonitor` easily since we already cache the workspace.

---------

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan commented on coderd/agentapi/connectionlog.go:
```diff
@@ -51,14 +53,31 @@ func (a *ConnLogAPI) ReportConnection(ctx context.Context, req *agentproto.Repor
 		}
 	}
 
+	// Inject RBAC object into context for dbauthz fast path, avoid having to
```
I merged the change with the helper, and used tasks to have AI update the other call sites for context injection in the latest commit. LMK what you guys think @zedkipp @Emyrk , it essentially just makes the error handling more consistent, we really only reduced LOC by about 10?
